File: examples/webauthn_large_blob.py
"""
WebAuthn largeBlob extension – read and write large blobs.

Demonstrates:
  1. Creating a credential with largeBlob support required
  2. Writing arbitrary data to the large blob associated with the credential
  3. Reading the data back in a subsequent authentication

Unlike credBlob (which is limited to ~32 bytes), largeBlob can store
much larger data using the authenticator's dedicated blob storage.

Usage: uv run python examples/webauthn_large_blob.py

⚠ This example creates a credential and writes to large blob storage.
"""

from __future__ import annotations

import base64
import getpass
import json
import os
import sys

from yubikit.core.fido import FidoConnection
from yubikit.core.smartcard import SmartCardConnection
from yubikit.device import list_all_devices
from yubikit.webauthn import ClientDataCollector, UserInteraction, WebAuthnClient

ORIGIN = "https://example.com"
RP_ID = "example.com"


class ConsoleInteraction(UserInteraction):
    def prompt_up(self) -> None:
        print("\n👆 Touch your security key...")

    def request_pin(self, permissions: int, rp_id: str | None) -> str | None:
        try:
            pin = getpass.getpass("🔑 Enter PIN: ")
            return pin if pin else None
        except (EOFError, KeyboardInterrupt):
            return None

    def request_uv(self, permissions: int, rp_id: str | None) -> bool:
        print("🔒 Biometric verification requested - proceeding")
        return True


class SimpleCollector(ClientDataCollector):
    def collect_create(self, options_json: str) -> tuple[bytes, str]:
        options = json.loads(options_json)
        rp_id = options.get("rp", {}).get("id", RP_ID)
        client_data = json.dumps(
            {
                "type": "webauthn.create",
                "challenge": options["challenge"],
                "origin": ORIGIN,
                "crossOrigin": False,
            },
            separators=(",", ":"),
        ).encode()
        return client_data, rp_id

    def collect_get(self, options_json: str) -> tuple[bytes, str]:
        options = json.loads(options_json)
        rp_id = options.get("rpId", RP_ID)
        client_data = json.dumps(
            {
                "type": "webauthn.get",
                "challenge": options["challenge"],
                "origin": ORIGIN,
                "crossOrigin": False,
            },
            separators=(",", ":"),
        ).encode()
        return client_data, rp_id


def b64url(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


def main() -> None:
    devices = list_all_devices()
    if not devices:
        print("No YubiKeys found.", file=sys.stderr)
        sys.exit(1)

    dev, info = devices[0]
    print(f"Using: {info.serial or 'Unknown serial'}")

    conn: FidoConnection | SmartCardConnection
    if dev.supports_connection(FidoConnection):  # type: ignore[arg-type]
        conn = dev.open_connection(FidoConnection)  # type: ignore[arg-type]
    elif dev.supports_connection(SmartCardConnection):
        conn = dev.open_connection(SmartCardConnection)
    else:
        print("No usable connection.", file=sys.stderr)
        sys.exit(1)

    client = WebAuthnClient(conn, ConsoleInteraction(), SimpleCollector())

    # -- Registration with largeBlob support required --
    # largeBlob requires a resident key (discoverable credential)
    print("\n━━━ Registration (largeBlob: required) ━━━")

    reg_json = client.make_credential(
        json.dumps(
            {
                "rp": {"name": "LargeBlob Example", "id": RP_ID},
                "user": {
                    "id": b64url(b"largeblob-user"),
                    "name": "largeblob@example.com",
                    "displayName": "LargeBlob User",
                },
                "challenge": b64url(os.urandom(32)),
                "pubKeyCredParams": [{"type": "public-key", "alg": -7}],
                "timeout": 60000,
                "authenticatorSelection": {
                    "residentKey": "required",
                },
                "extensions": {
                    "largeBlob": {"support": "required"},
                },
            }
        )
    )

    reg = json.loads(reg_json)
    cred_id = reg["id"]
    print(f"✅ Credential: {cred_id}")

    ext = reg.get("clientExtensionResults", {})
    lb = ext.get("largeBlob", {})
    if lb:
        print(f"  largeBlob supported: {lb.get('supported')}")

    # -- Write a blob --
    blob_data = (
        b"This is a larger piece of data stored via the largeBlob extension. "
        b"It can hold more than the 32 bytes credBlob is limited to."
    )
    print(f"\n━━━ Authentication: write blob ({len(blob_data)} bytes) ━━━")

    assertions = client.get_assertion(
        json.dumps(
            {
                "challenge": b64url(os.urandom(32)),
                "timeout": 60000,
                "rpId": RP_ID,
                "allowCredentials": [{"type": "public-key", "id": cred_id}],
                "userVerification": "discouraged",
                "extensions": {
                    "largeBlob": {"write": b64url(blob_data)},
                },
            }
        )
    )

    a = json.loads(assertions[0])
    ext = a.get("clientExtensionResults", {})
    lb = ext.get("largeBlob", {})
    if lb:
        print(f"✅ Blob written: {lb.get('written')}")

    # -- Read the blob back --
    print("\n━━━ Authentication: read blob ━━━")

    assertions = client.get_assertion(
        json.dumps(
            {
                "challenge": b64url(os.urandom(32)),
                "timeout": 60000,
                "rpId": RP_ID,
                "allowCredentials": [{"type": "public-key", "id": cred_id}],
                "userVerification": "discouraged",
                "extensions": {
                    "largeBlob": {"read": True},
                },
            }
        )
    )

    a = json.loads(assertions[0])
    ext = a.get("clientExtensionResults", {})
    lb = ext.get("largeBlob", {})
    if lb and lb.get("blob"):
        data = base64.urlsafe_b64decode(lb["blob"] + "==")
        print(f"✅ Read {len(data)} bytes")
        try:
            print(f'  Content: "{data.decode()}"')
        except UnicodeDecodeError:
            print(f"  Content: {data.hex()}")
        if data == blob_data:
            print("  ✅ Data matches what was written!")
        else:
            print("  ❌ Data does not match!")
    else:
        print("  (no blob data returned)")

    client.close()
    print("\nDone.")


if __name__ == "__main__":
    main()

File: crates/yubikit/examples/webauthn_large_blob.rs
// Example: WebAuthn largeBlob extension – read and write large blobs.
//
// Demonstrates:
//   1. Creating a credential with `largeBlob.support = "required"`
//   2. Writing arbitrary data to the large blob associated with the credential
//   3. Reading the data back in a subsequent authentication
//
// Unlike credBlob (which is limited to ~32 bytes), largeBlob can store
// much larger data using the authenticator's dedicated blob storage.
//
// Usage: cargo run -p yubikit --example webauthn_large_blob
//
// ⚠ This example creates a credential and writes to large blob storage.

use std::io::{self, Write};

use yubikit::ctap::CtapSession;
use yubikit::ctap2::{Ctap2Session, Permissions};
use yubikit::transport::ctaphid::{HidFidoConnection, list_fido_devices};
use yubikit::webauthn::extensions::{
    AuthenticationExtensionInputs, RegistrationExtensionInputs,
    large_blob::{AuthenticationInput, LargeBlobSupport, RegistrationInput},
};
use yubikit::webauthn::{
    AuthenticatorSelectionCriteria, ClientDataCollector, CollectedClientData,
    PublicKeyCredentialCreationOptions, PublicKeyCredentialDescriptor,
    PublicKeyCredentialParameters, PublicKeyCredentialRequestOptions, PublicKeyCredentialRpEntity,
    PublicKeyCredentialType, PublicKeyCredentialUserEntity, ResidentKeyRequirement,
    UserInteraction, UserVerificationRequirement, WebAuthnClient,
};

// ---------------------------------------------------------------------------
// Shared helpers
// ---------------------------------------------------------------------------

struct ConsoleInteraction;

impl UserInteraction for ConsoleInteraction {
    fn prompt_up(&self) {
        println!("\n👆 Touch your security key...");
    }
    fn request_pin(&self, _permissions: Permissions, _rp_id: Option<&str>) -> Option<String> {
        print!("🔑 Enter PIN: ");
        io::stdout().flush().ok();
        let mut pin = String::new();
        io::stdin().read_line(&mut pin).ok()?;
        let pin = pin.trim().to_string();
        if pin.is_empty() { None } else { Some(pin) }
    }
    fn request_uv(&self, _p: Permissions, _rp: Option<&str>) -> bool {
        true
    }
}

const ORIGIN: &str = "https://example.com";
const RP_ID: &str = "example.com";

struct SimpleCollector;

impl ClientDataCollector for SimpleCollector {
    fn collect_create(
        &self,
        options: &PublicKeyCredentialCreationOptions,
    ) -> Result<(CollectedClientData, String), String> {
        let rp_id = options.rp.id.clone().unwrap_or_else(|| RP_ID.to_string());
        Ok((
            CollectedClientData::create("webauthn.create", &options.challenge, ORIGIN, false),
            rp_id,
        ))
    }
    fn collect_get(
        &self,
        options: &PublicKeyCredentialRequestOptions,
    ) -> Result<(CollectedClientData, String), String> {
        let rp_id = options.rp_id.clone().unwrap_or_else(|| RP_ID.to_string());
        Ok((
            CollectedClientData::create("webauthn.get", &options.challenge, ORIGIN, false),
            rp_id,
        ))
    }
}

fn random_challenge() -> Vec<u8> {
    use std::collections::hash_map::DefaultHasher;
    use std::hash::{Hash, Hasher};
    use std::time::SystemTime;
    let mut h = DefaultHasher::new();
    SystemTime::now().hash(&mut h);
    std::process::id().hash(&mut h);
    let a = h.finish().to_le_bytes();
    let mut h2 = DefaultHasher::new();
    a.hash(&mut h2);
    42u64.hash(&mut h2);
    let b = h2.finish().to_le_bytes();
    [a, b, a, b].concat()
}

fn hex(bytes: &[u8]) -> String {
    bytes.iter().map(|b| format!("{b:02x}")).collect()
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

fn main() {
    let devices = list_fido_devices().expect("failed to list HID devices");
    let dev = devices.first().unwrap_or_else(|| {
        eprintln!("No FIDO HID devices found.");
        std::process::exit(1);
    });

    println!("Using device: {} (PID=0x{:04X})", dev.path, dev.pid);

    let conn = HidFidoConnection::open(dev).expect("failed to open HID connection");
    let ctap = match CtapSession::new_fido(conn) {
        Ok(s) => s,
        Err((e, _)) => {
            eprintln!("CTAP session error: {e}");
            std::process::exit(1);
        }
    };
    let session = Ctap2Session::new(ctap).expect("CTAP2 init failed");

    let info = session.info();
    if info.options.get("largeBlobs") != Some(&true) {
        eprintln!("Authenticator does not support largeBlobs.");
        std::process::exit(1);
    }
    println!("✓ largeBlobs supported");

    let mut client = WebAuthnClient::new(session, ConsoleInteraction, SimpleCollector);

    // -- Registration with largeBlob support required --
    // largeBlob requires a resident key (discoverable credential)
    println!("\n━━━ Registration (largeBlob: required) ━━━");

    let create_options = PublicKeyCredentialCreationOptions {
        rp: PublicKeyCredentialRpEntity {
            name: "LargeBlob Example".to_string(),
            id: Some(RP_ID.to_string()),
        },
        user: PublicKeyCredentialUserEntity {
            id: b"largeblob-user".to_vec(),
            name: Some("largeblob@example.com".to_string()),
            display_name: Some("LargeBlob User".to_string()),
        },
        challenge: random_challenge(),
        pub_key_cred_params: vec![PublicKeyCredentialParameters {
            type_: PublicKeyCredentialType::PublicKey,
            alg: -7,
        }],
        timeout: Some(60_000),
        exclude_credentials: None,
        authenticator_selection: Some(AuthenticatorSelectionCriteria {
            authenticator_attachment: None,
            resident_key: Some(ResidentKeyRequirement::Required),
            user_verification: None,
        }),
        hints: None,
        attestation: None,
        attestation_formats: None,
        extensions: Some(RegistrationExtensionInputs {
            large_blob: Some(RegistrationInput {
                support: LargeBlobSupport::Required,
            }),
            ..Default::default()
        }),
    };

    let reg = match client.make_credential(&create_options) {
        Ok(r) => r,
        Err(e) => {
            eprintln!("Registration failed: {e}");
            std::process::exit(1);
        }
    };

    println!("✅ Credential: {}", hex(&reg.id));
    if let Some(ref ext) = reg.client_extension_results
        && let Some(ref lb) = ext.large_blob
    {
        println!("  largeBlob supported: {}", lb.supported);
    }

    // -- Write a blob --
    let blob_data = b"This is a larger piece of data stored via the largeBlob extension. \
        It can hold more than the 32 bytes credBlob is limited to."
        .to_vec();
    println!(
        "\n━━━ Authentication: write blob ({} bytes) ━━━",
        blob_data.len()
    );

    let write_options = PublicKeyCredentialRequestOptions {
        challenge: random_challenge(),
        timeout: Some(60_000),
        rp_id: Some(RP_ID.to_string()),
        allow_credentials: Some(vec![PublicKeyCredentialDescriptor {
            type_: PublicKeyCredentialType::PublicKey,
            id: reg.id.clone(),
            transports: None,
        }]),
        user_verification: Some(UserVerificationRequirement::Discouraged),
        hints: None,
        extensions: Some(AuthenticationExtensionInputs {
            large_blob: Some(AuthenticationInput::write(blob_data.clone())),
            ..Default::default()
        }),
    };

    let assertions = match client.get_assertion(&write_options) {
        Ok(a) => a,
        Err(e) => {
            eprintln!("Write failed: {e}");
            std::process::exit(1);
        }
    };

    if let Some(ref ext) = assertions[0].client_extension_results
        && let Some(ref lb) = ext.large_blob
    {
        println!("✅ Blob written: {:?}", lb.written);
    }

    // -- Read the blob back --
    println!("\n━━━ Authentication: read blob ━━━");

    let read_options = PublicKeyCredentialRequestOptions {
        challenge: random_challenge(),
        timeout: Some(60_000),
        rp_id: Some(RP_ID.to_string()),
        allow_credentials: Some(vec![PublicKeyCredentialDescriptor {
            type_: PublicKeyCredentialType::PublicKey,
            id: reg.id.clone(),
            transports: None,
        }]),
        user_verification: Some(UserVerificationRequirement::Discouraged),
        hints: None,
        extensions: Some(AuthenticationExtensionInputs {
            large_blob: Some(AuthenticationInput::read()),
            ..Default::default()
        }),
    };

    let assertions = match client.get_assertion(&read_options) {
        Ok(a) => a,
        Err(e) => {
            eprintln!("Read failed: {e}");
            std::process::exit(1);
        }
    };

    if let Some(ref ext) = assertions[0].client_extension_results
        && let Some(ref lb) = ext.large_blob
    {
        if let Some(ref data) = lb.blob {
            println!("✅ Read {} bytes", data.len());
            match std::str::from_utf8(data) {
                Ok(s) => println!("  Content: \"{s}\""),
                Err(_) => println!("  Content: {}", hex(data)),
            }
            if *data == blob_data {
                println!("  ✅ Data matches what was written!");
            } else {
                println!("  ❌ Data does not match!");
            }
        } else {
            println!("  (no blob data returned)");
        }
    }

    let _session = client.into_session();
    println!("\nDone.");
}
